Coin control in CoinJoin account

komret opened the pull request:
## Description

The first two commits are just a minor fix and refactoring. Following commits:

- add new UTXO categories. Standard accounts still use the two categories (default + dust), while the CoinJoin account has three (private + not private + dust). Note that each category is only displayed when there are any UTXOs belonging to it. I modified behaviour of the check-all checkbox at the top so that it always selects the top category (e.g. will select all non-private UTXOs if there are no private ones. Note that dust is not recognised on Regtest as there is no `dustLimit`.
- add new UTXO tags (icons with tooltip). Apart from change address, there are now two new tags for UTXOs registered for Coinjoin and those not available for CoinJoin (too small or blocked by coordinator - the latter not yet implemented).
- add `UtxoAnonymity` component displayed in coin control and transaction detail. The same icon was used in `AnonymityLevelIndicator` to unify it. In coin control, this is displayed right from the timestamp because it has variable length.

## Related Issue

Resolve #6215

## Screenshots (if appropriate):

_Not private_ category and a UTXO with a _blocked_ tag and _1 of 1_ anonymity:
![Screenshot 2022-10-31 at 17 13 02](https://user-images.githubusercontent.com/42465546/199056515-d9c46c7b-7b5a-4e4b-8c20-7c48d0251dd2.png)
Transaction detail showing the anonymity of my UTXO:
![Screenshot 2022-10-31 at 17 13 22](https://user-images.githubusercontent.com/42465546/199056933-de41effc-091a-4db1-94be-aaf738ba41b1.png)
New icon in `AnonymityLevelIndicator`:
![Screenshot 2022-10-31 at 17 20 02](https://user-images.githubusercontent.com/42465546/199057549-e3a98ba1-682a-4235-a2e0-236b9eb66409.png)

matejkriz commented on packages/suite/src/components/wallet/CoinControl/UtxoSelection.tsx:
```diff
@@ -123,14 +125,21 @@ interface UtxoSelectionProps {
+        getUtxoOutpoint(utxo),
+    );
+    const isUnavailableForCoinjoin =
+        account.accountType === 'coinjoin' && Number(utxo.amount) < 5000; // TODO: should include UTXOs banned from CoinJoin as well
```
maybe BigNumber for utxo.amount? Why 5000? I don't understand this condition.

komret replied:
Thanks, 6433f762959dbbd57722c197fdbdb73e7c013b61 explains this. It is WIP.

matejkriz reviewed (approved):
👍 it looks good

dahaca commented on packages/suite/src/hooks/wallet/form/useUtxoSelection.ts:
```diff
@@ -40,19 +39,30 @@ export const useUtxoSelection = ({
+    const lowAnonymityUtxos: AccountUtxo[] = [];
+    const dustUtxos: AccountUtxo[] = [];
+    account?.utxo?.forEach(utxo => {
+        switch (excludedUtxos[getUtxoOutpoint(utxo)]) {
```
nit: perhaps rename to `excludedUtxoOutpoints?`

komret replied:
I would keep this as agreed previously with @szymonlesisz. The outpoints are just pointers to UTXOs.

dahaca commented on packages/suite/src/hooks/wallet/form/useUtxoSelection.ts:
```diff
@@ -95,7 +105,17 @@ export const useUtxoSelection = ({
-            setValue('selectedUtxos', [...dustUtxos.filter(isSelected), ...spendableUtxos]);
+            setValue(
+                'selectedUtxos',
+                [
```
nit: put that in a constant beforehand for readability

komret replied:
I refactored it so that it is more readable: daffb4d07e4e1abed98bd1d2733410a2874c55a5

dahaca commented on packages/suite/src/hooks/wallet/useSendForm.ts:
```diff
@@ -65,21 +65,17 @@ const getStateFromProps = (props: UseSendFormProps) => {
-            if (coinjoinSession && coinjoinSession.registeredUtxos.includes(outpoint)) {
-                // utxo is registered in coinjoin
-                excludedUtxos[outpoint] = 'mixing';
+            if (new BigNumber(utxo.amount).lte(Number(coinFees.dustLimit))) {
```
the change to `lte` is deliberate?

komret replied:
Yes, to correspond with the [validation](https://github.com/trezor/trezor-suite/blob/develop/packages/connect/src/api/signTransaction.ts#L103). It is quite unintuitive, though.

dahaca reviewed (commented):
1. Perhaps a mention of coin control should be added to the orange warning, since it's present in the tooltip <img width="775" alt="image" src="https://user-images.githubusercontent.com/45338719/199269497-40b1c53d-b961-45b0-be32-2298360525b1.png">
2. Should we have `Send not anonymized Coins` on the send button [like on the design](https://www.figma.com/file/hZ22ckyTRlCYRLTNUbTdWO/Exploring-Accounts?node-id=2195%3A36491) or keep `Review & Send`? @hynek-jina 
<img width="318" alt="image" src="https://user-images.githubusercontent.com/45338719/199269637-fa9925fe-043b-4771-827c-37fb097f69f4.png">

hynek-jina commented:
Nice catch @dahaca 
1) Feel free to extend it to `Either anonymize more funds, lower the anonymity level, or select coins manually at your own risk`
2) I would stick to designed `Send not anonymized Coins`

komret commented:
I changed the messages in c33696db7251c548814ba7309dbb04ecec077645.

Send button:
![Screenshot 2022-11-01 at 19 31 57](https://user-images.githubusercontent.com/42465546/199313367-03f8531b-d57d-48d6-a9bd-50bf518b526f.png)
I added a new message used when not anonymized UTXO is selected and broadcast is off:
![Screenshot 2022-11-01 at 19 32 07](https://user-images.githubusercontent.com/42465546/199313550-2a19f0be-d2b6-4145-8540-c339ceb132d5.png)
I removed the line break from the orange warning so that it still fits on two lines:
![Screenshot 2022-11-01 at 19 34 34](https://user-images.githubusercontent.com/42465546/199313329-7b3d1d2b-a34c-465a-87da-c717b5f35901.png)

I also noticed missing `margin` on `Pagination` and fixed it.

hynek-jina commented:
> I changed the messages in [c33696d](https://github.com/trezor/trezor-suite/commit/c33696db7251c548814ba7309dbb04ecec077645).
> 
> Send button: ![Screenshot 2022-11-01 at 19 31 57](https://user-images.githubusercontent.com/42465546/199313367-03f8531b-d57d-48d6-a9bd-50bf518b526f.png) I added a new message used when not anonymized UTXO is selected and broadcast is off: ![Screenshot 2022-11-01 at 19 32 07](https://user-images.githubusercontent.com/42465546/199313550-2a19f0be-d2b6-4145-8540-c339ceb132d5.png) I removed the line break from the orange warning so that it still fits on two lines: ![Screenshot 2022-11-01 at 19 34 34](https://user-images.githubusercontent.com/42465546/199313329-7b3d1d2b-a34c-465a-87da-c717b5f35901.png)
> 
> I also noticed missing `margin` on `Pagination` and fixed it.

It's a small detail.. but if that fits I would move the second sentence on the second line and add `.` dot at the end.. But I get It can be changed in Crowdin, right?

komret commented:
@hynek-jina, I added the period. The sentence does not fit on one line though and it might not fit in other languages either.

dahaca reviewed (approved):
🙌
